mysql-plugin/src/main/java/io/cdap/plugin/mysql/MysqlConnector.java

```diff
@@ -87,17 +101,51 @@ protected void setConnectorSpec(ConnectorSpecRequest request, DBConnectorPath pa
   }
 
   @Override
-  protected String getTableQuery(String database, String schema, String table) {
-    return String.format("SELECT * FROM `%s`.`%s`", database, table);
+  public StructuredRecord transform(LongWritable longWritable, MysqlDBRecord record) {
+    return record.getRecord();
+  }
+
+  @Override
+  protected String getTableName(String database, String schema, String table) {
+    return String.format("`%s`.`%s`", database, table);
   }
 
   @Override
-  protected String getTableQuery(String database, String schema, String table, int limit) {
-    return String.format("SELECT * FROM `%s`.`%s` LIMIT %d", database, table, limit);
+  protected String getRandomQuery(String tableName, int limit) {
+      // This query doesn't guarantee exactly "limit" number of rows
+      // Note that we input "limit" with a trailing zero so that division gives an exact result
+      return String.format("SELECT * FROM %s\n" +
+                             "WHERE rand() < %d.0 / (SELECT COUNT(*) FROM %s)",
+                           tableName, limit, tableName);
   }
 
   @Override
-  public StructuredRecord transform(LongWritable longWritable, MysqlDBRecord mysqlDBRecord) {
-    return mysqlDBRecord.getRecord();
+  protected String getStratifiedQuery(String tableName, int limit, String strata) {
+    return String.format("WITH t_%s AS (\n" +
+                           "    SELECT *,\n" +
+                           "    ROW_NUMBER() OVER (ORDER BY %s, RAND()) AS sqn_%s,\n" +
+                           "    COUNT(*) OVER () AS c_%s\n" +
+                           "    FROM %s\n" +
+                           "  )\n" +
+                           "SELECT * FROM t_%s\n" +
+                           "WHERE MOD(sqn_%s, CAST(c_%s / %d AS UNSIGNED)) = 1\n" +
+                           "ORDER BY %s\n" +
+                           "LIMIT %d",
+                         sessionID, strata, sessionID, sessionID, tableName, sessionID, sessionID, sessionID,
+                         limit, strata, limit);
+  }
+
+  @Override
+  protected Schema loadTableSchema(Connection connection, String query, @Nullable Integer timeoutSec)
+    throws SQLException {
+    Statement statement = connection.createStatement();
+    statement.setMaxRows(1);
+    if (timeoutSec != null) {
+      statement.setQueryTimeout(timeoutSec);
+    }
+    ResultSet resultSet = statement.executeQuery(query);
+    MysqlSchemaReader schemaReader = (MysqlSchemaReader) getSchemaReader();
+    schemaReader.setSessionID(sessionID);
```

Comment:
if you set the sessionId in the getSchemaReader() method, you don't need to do a cast above. By removing the cast, it also seems like it would also allow all the logic to be pulled up to the parent class instead of duplicated in each connector

Reply:
Changed. Thanks!